pkg/mailboxwatch/impl.go

```diff
@@ -0,0 +1,200 @@
+/*
+Copyright 2023 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package mailboxwatch
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+
+	machmeta "k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	k8ssets "k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/watch"
+	upstreamcache "k8s.io/client-go/tools/cache"
+
+	tenancyv1a1 "github.com/kcp-dev/kcp/pkg/apis/tenancy/v1alpha1"
+	tenancyv1a1informers "github.com/kcp-dev/kcp/pkg/client/informers/externalversions/tenancy/v1alpha1"
+	"github.com/kcp-dev/logicalcluster/v3"
+
+	"github.com/kcp-dev/edge-mc/pkg/placement"
+)
+
+func newFilteredListerWatcher(
+	mailboxWorkspacePreInformer tenancyv1a1informers.WorkspaceInformer,
+	unfiltered upstreamcache.ListerWatcher,
+) *filteredListerWatcher {
+	flw := &filteredListerWatcher{
+		unfiltered: unfiltered,
+		clusters:   k8ssets.NewString(),
+	}
+	mailboxWorkspacePreInformer.Informer().AddEventHandler(flw)
+	return flw
+}
+
+type filteredListerWatcher struct {
+	sync.Mutex
+	unfiltered upstreamcache.ListerWatcher
+	clusters   k8ssets.String
+}
+
+func (flw *filteredListerWatcher) OnAdd(obj any) {
+	flw.setInclusion(obj, true)
+}
+
+func (flw *filteredListerWatcher) OnUpdate(old, obj any) {
+	flw.setInclusion(obj, true)
+}
+
+func (flw *filteredListerWatcher) OnDelete(obj any) {
+	flw.setInclusion(obj, false)
+}
+
+func (flw *filteredListerWatcher) setInclusion(obj any, include bool) {
+	ws := obj.(*tenancyv1a1.Workspace)
+	mbwsName := ws.Name
+	mbwsNameParts := strings.Split(mbwsName, placement.WSNameSep)
+	if len(mbwsNameParts) != 2 {
+		// Only accept the workspace if its name looks like a mailbox workspace name
+		include = false
+	}
+	cluster := ws.Spec.Cluster
+	flw.Lock()
+	defer flw.Unlock()
+	if include {
+		flw.clusters.Insert(cluster)
+	} else {
+		flw.clusters.Delete(cluster)
+	}
+}
+
+type myList struct {
+	metav1.TypeMeta
+	metav1.ListMeta
+	items []runtime.Object
```

Comment:
Is it OK with `items`?
I get the following error (`*mailboxwatch.myList is not a list: no Items field in this object`).
```
E0329 09:40:59.750244   51001 reflector.go:138] k8s.io/client-go@v0.0.0-20230210192259-aaa28aa88b2d/tools/cache/reflector.go:215: Failed to watch *v1.ConfigMap: unable to understand list result &mailboxwatch.myList{TypeMeta:v1.TypeMeta{Kind:"", APIVersion:""}, ListMeta:v1.ListMeta{SelfLink:"", ResourceVersion:"4001", Continue:"", RemainingItemCount:(*int64)(nil)}, items:[]runtime.Object{(*v1.ConfigMap)(0x140003aaf00)}} (*mailboxwatch.myList is not a list: no Items field in this object)
```

If I change it to `Items`, the error is gone and the informer successfully notifies the handler of the Add/Update/Delete event.

Reply:
This is the case when I use `"k8s.io/client-go/kubernetes"` for `k8sClientset`. Originally, I am used to use `"github.com/kcp-dev/client-go/kubernetes"`. I am still doing try-and-error for https://github.com/kcp-dev/edge-mc/pull/265#issuecomment-1487763749. I am trying non-cluster client.

The full code is https://gist.github.com/yana1205/b9303ea12f3e0230e17677d93fc1ab78#file-main-with-non-cluster-cliientset-go

Reply:
Oh, dang, I meant to correct the capitalization of `items`, I must have forgotten it.

Reply:
For authorization problems you also need to look at the user involved.  I do not remember whether it shows up on the client, shows up on the server, but do look for it.

For the informer on mailbox workspaces, it should be focused on one specific logical cluster --- the edge service provider workspace.

The clientset-from-config function that you call has to match the server URL in the client config.  Call the all-cluster clientset constructor when the server URL is short (does not include `/clusters/$something_even_star`), call the single-cluster-focused clientset constructor when the server URL is long (includes `/clusters/$something_not_star`).  As far as I know, it should be valid to either make an all-cluster clientset and then use a `Cluster(.)` method to scope it down or start with a single-cluster clientset.

Reply:
What is the relationship between the gist you referenced and the github content you referenced?

Reply:
I think that the C2P Controller will need at least two kube client configs, just as the placement translator does.  See https://docs.kcp-edge.io/docs/coding-milestones/poc2023q1/placement-translator/#usage .  One kube client config points specifically at the edge service provider workspace and authorizes reading the Workspace objects from there. The other kube client config points at the server base and authorizes everything, including reading policy reports from all clusters.

You might want to re-use the code at https://github.com/kcp-dev/edge-mc/blob/a2f67fc36838e6052043896995a16441ba30527a/cmd/placement-translator/main.go#L65-L92 for command line flags for those two kube client configs.

Reply:
The latest revision fixes the capitalization of Items.